modules/javafx.media/src/main/java/com/sun/media/jfxmedia/locator/HLSConnectionHolder.java

```diff
@@ -257,811 +413,1192 @@ private ReadableByteChannel openHeaderChannel() throws IOException {
                 playlist.update(currentPlaylist.getNextMediaFile());
-                playlistThread.setReloadPlaylist(playlist);
+                playlistLoader.setReloadPlaylist(playlist);
             }
 
             playlist.setForceDiscontinuity(true);
+            // Copy index when switching playlist, so we continue reading
+            // from correct position.
+            playlist.setMediaFileIndex(currentPlaylist.getMediaFileIndex());
             currentPlaylist = playlist;
             if (currentPlaylist.isFragmentedMP4()) {
                 sendHeader = true;
             }
+
+            // We switched playlist and we need to check if we need to
+            // switch audio stream as well.
+            if (getAudioStream() != null) {
+                String audioGroupID = currentPlaylist.getAudioGroupID();
+                Playlist audioPlaylist = variantPlaylist.getAudioExtPlaylist(audioGroupID);
+                getAudioStream().setNewCurrentPlaylist(audioPlaylist);
+            }
+        }
+    }
+
+    private void adjustBitrateAudioExt() {
+        // If video stream provided new playlist, then use it.
+        synchronized(newPlaylistLock) {
+            if (newCurrentPlaylist != null && newCurrentPlaylist != currentPlaylist) {
+                if (currentPlaylist.isLive()) {
+                    newCurrentPlaylist.update(currentPlaylist.getNextMediaFile());
+                    playlistLoader.setReloadAudioExtPlaylist(newCurrentPlaylist);
+                }
+
+                newCurrentPlaylist.setForceDiscontinuity(true);
+                // Copy index when switching playlist, so we continue reading
+                // from correct position.
+                newCurrentPlaylist.setMediaFileIndex(currentPlaylist.getMediaFileIndex());
+                currentPlaylist = newCurrentPlaylist;
+                if (currentPlaylist.isFragmentedMP4()) {
+                    sendHeader = true;
+                }
+
+                newCurrentPlaylist = null;
+            }
         }
     }
 
-    private static String stripParameters(String mediaFile) {
+    static String stripParameters(String mediaFile) {
         int qp = mediaFile.indexOf('?');
         if (qp > 0) {
             mediaFile = mediaFile.substring(0, qp); // Strip all possible http parameters.
         }
         return mediaFile;
     }
 
-    private class PlaylistThread extends Thread {
+    Playlist getCurrentPlaylist() {
+        return currentPlaylist;
+    }
+}
 
-        public static final int STATE_INIT = 0;
-        public static final int STATE_EXIT = 1;
-        public static final int STATE_RELOAD_PLAYLIST = 2;
-        private final BlockingQueue<Integer> stateQueue = new LinkedBlockingQueue<>();
-        private URI playlistURI = null;
-        private Playlist reloadPlaylist = null;
-        private final Object reloadLock = new Object();
-        private volatile boolean stopped = false;
+class PlaylistLoader extends Thread {
```

Comment:
When these classes were nested classes I got issues with second instance of HLSConnectionHolder. If I remember correctly nested classes of second instance of HLSConnectionHolder were using fields of first HLSConnectionHolder instance. Maybe because I initiated second instance incorrectly. To avoid any such potential issues I decided to move away from nested classes. I would prefer to keep as is or better to move all nested classes under separate package (com.sun.media.jfxmedia.locator.hls).

Reply:
Did you declare them as `static` (nested) classes? If not, then yes, they will have the behavior you mentioned. A non-static "inner" class exists within an instance of the outer class. A static "nested" class does not. Other than scoping, a nested class behaves like a top-level class.

If you do want to keep them as separate top-level classes, then please move each to its own file. I would not recommend creating a new package, though, since that will involve more changes to make the classes and elements you need public.

Reply:
No, I did not declared them as `static`. Fixed by declaring these classes as nested static classes.